RSBR Sensor- In Progress (#284)
monoDriveIO/documentation: docs/monoDrive_home/RSBR.md

# RSBR  

> RSBR Description

Reviewer: ```suggestion
The Radar Shooting Bouncing Ray (RSBR) sensor provides ground truth hit point access to targets that are within the field-of-view of a simulated radar. The output of the sensor provides target intersection information at the resolution and field-of-view specified by the configuration. 
```

## Configuration

``` json
{
    "type": "RSBR",
    "listen_port": 0,
    "location": {
        "x": 0.0,
        "y": 0.0,
        "z": 0.0
    },
    "rotation": {
        "pitch": 0.0,
        "roll": 0.0,
        "yaw": 0.0
    },
    "azi_fov": 90,
    "ele_fov": 90,
    "scan_distance": 3000,
    "debug_draw": false,
    "num_threads": 8,
    "resolution": 5,
    "max_returns": 100000,
    "raycasts_per_thread": 100000,

>     "max_targets_capture": 1000

Reviewer: Remove this.

}
```

>  - **azi_fov:** The angle in the x-y plane to detect objects in degrees.

Reviewer: ```suggestion
 - **azi_fov:** The horizontal field-of-view along the azimuth of the sensor in degrees.
```

>  - **ele_fov:** The angle in the y-z plane to detect objects in degrees.

Reviewer: ```suggestion
 - **ele_fov:** The vertical field-of-view along the elevation of the sensor in degrees.
```

>  - **scan_distance:** The maximum distance the sensor will detect objects in meters.

Reviewer: ```suggestion
 - **scan_distance:** The maximum distance the sensor will detect objects in centimeters.
```

>  - **debug_draw:** If true, the oriented bounding boxes for each actor in the output will be drawn.

Reviewer: ```suggestion
 - **debug_draw:** If true, the rays that are cast will be drawn in editor mode.
```

>  - **num_threads:** Description.

Reviewer: ```suggestion
 - **num_threads:** The total number of threads that will be used to compute the ray casts.
```

>  - **resolution:** Description.

Reviewer: ```suggestion
 - **resolution:** The spacing between individual rays in centimeters.
```

>  - **max_returns:** Description.

Reviewer: ```suggestion
 - **max_returns:** The maximum number of targets that will be returned by the sensor.
```

>  - **raycasts_per_thread:** Description.

Reviewer: ```suggestion
 - **raycasts_per_thread:** The total number of rays that will be cast iteratively in each thread.
```

>  - **max_targets_capture:** Description.

Reviewer: This is not used. Just omit.

> ## Raw Output

Reviewer: The output is a raw buffer of `RSBRHitPoint`s. Here's the info for a single hit point: https://github.com/monoDriveIO/monodrive-client/blob/4d70b83b88bf2f1d6aeca3bb906a715b34883ffe/monodrive/core/src/DataFramePrimitives.h#L158

The first element of the buffer is the total number of hit points for the return: https://github.com/monoDriveIO/monodrive-client/blob/development/monodrive/core/src/DataFrame.cpp#L352

> Explanation

Reviewer: It would be good to include some output from the visualization. Maybe a screen shot from this video or the whole video: https://monodrive.slack.com/archives/C01662Y9D28/p1618406012126000
